### PR TITLE
Improved support for multiple modes

### PR DESCRIPTION
**Pull request by ben-spiller**
The biggest request I've received from PySys users is the ability to run a single test in multiple "modes" - for example against different web browsers, or databases, or whatever. 

There's already a concept of modes in PySys but a) by default when you invoke PySys it runs with mode='' even for a test that expects modes which makes test implementation a bit tricky and more importantly b) there is no way to run a test in multiple modes from a singel invocation of PySys, so you need to separately run PySys several times for each mode (perhaps manually specifying a different outdir for each one), and even then there is no way to aggregate results across them. 

In this PR we will retain the existing "modes" concept for compatibility purposes but add a project setting supportMultipleModesPerRun=true to toggle on a more powerful way of handling modes that adds a separate test descriptor to the job queue for each mode supported by each test. From a single invocation of "pysys run" you would be able to either run all tests against just their primary mode, or run all tests against all supported modes, or provide a list of mode include/excludes for more advanced cases. A suffix ("~MyModeHere") will be added automatically to test ids to keep them unique for reporting purposes, and outdir will also be kept separate. 

It's helpful to have a concept of one of the modes for each tests being the primary one, which for simplicity we define as the first one listed in the descriptor.

Command line looks like:
pysys run # runs all tests in their primary mode (including tests that have no modes)
pysys run -m ALL # ALL means run in all supported modes
pysys run -m Mode1,Mode2 # run tests in Mode1 and Mode2; ignores tests that do not support these modes or that have no mode
pysys run -m !Mode1,Mode2 # run tests in all modes (incl tests that have no mode) except Mode1 or Mode2
pysys run -m !PRIMARY # run tests in all modes except whichever is the primary mode (i.e. everything that's not included in a "pysys run" execution)

This involves special values for ALL and PRIMARY.  

For cases too complex to specify statically (e.g. if you want to add the same set of modes to multiple test directory trees, or want to decide them at runtime based on current OS etc), we introduce a pluggable DescriptorLoader which iterates over a specified directory and returns a list of modes. A typical customization would be to invoke the default implementation then add a dynamic set of extra modes to tests that have a particular group set - see PySys_internal_112 for an example. This API would also potentially allow users to programatically define descriptor wrappers for non-pysys testcases (e.g. JUnit Java classes and similar; see test _113 for a proof of concept), so it's quite powerful. 

Pysys print doesn't show copies of each id for all modes as that would be unhelpfully verbose, but you can use --full to see the available modes for each test if you want. 

Added inherit=false option to <modes> and <groups> in case someone wants to prevent parent modes being inherited rom pysysdirconfig.xml files. 

Also added an undocumented baserunner._testScheduler hook that cunningly gives a custom runner enough power to control what jobs are picked off the threadpool queue in what order - this will not yet be publicised to end-users, but allows us to experiment with creating a more advanced test scheduler for pysys over the next year and eventually contribute it back.

This PR also includes a reworking of the runOrderPriority concept on master; now it's called an execution order hint and there will be a way to globally specify deltas to be added to the individual descriptor hints based on mode and/or group. The default will be a larger order hint for non-primary modes, so all tests will execute in their primary mode before we start on extra modes. I believe this model provides enough power to express pretty sophisticated ordering both globally and at a per-test level. Only power users with big projects are likely to be using this level of functionality. 

I think it's all in pretty good shape now, and I've added content to the changelog and userguide to ensure end-users  are clear on the mode. Haven't yet heard any feedback on this draft PR. Since overall this PR is one of the biggest additions to the PySys model recently it'd be good to confirm you guys are happy with it. :)

### CODE REVIEW

File: pysys/baserunner.py [hunk 1]
```diff
@@ -100,7 +104,8 @@ def __init__(self, record, purge, cycle, mode, threads, outsubdir, descriptors,
 		@param record: Indicates if the test results should be recorded 
 		@param purge: Indicates if the output subdirectory should be purged on C{PASSED} result
 		@param cycle: The number of times to execute the set of requested testcases
-		@param mode: The user defined mode to run the testcases in
+		@param mode: Only used if supportMultipleModesPerRun=False; specifies the single mode 
+		tests will be run with. 
 		@param threads: The number of worker threads to execute the requested testcases
 		@param outsubdir: The name of the output subdirectory
 		@param descriptors: List of XML descriptor containers detailing the set of testcases to be run
```
Comment (rsm31) [inline pysys/baserunner.py]: These are no longer specifically XML descriptors according to the changed comment above.

File: pysys-examples/internal/testcases/PySys_internal_104/Input/dir1/Test_RootNoInheritDescriptor/pysystest.xml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  <description> 
+    <title>Nested testcase</title>    
+    <purpose><![CDATA[Nested purpose
+
+]]>
+    </purpose>
+  </description>
+
+  <classification>
+    <groups inherit="fAlse">
```
[the remark this answers is not in the document]
Comment (ben-spiller) [reply pysys-examples/internal/testcases/PySys_internal_104/Input/dir1/Test_RootNoInheritDescriptor/pysystest.xml]: nope that's me deliberately testing that we do a case-insensitive check on the boolean :)

File: pysys/baserunner.py (hunk 1 above)
Comment (ben-spiller) [reply pysys/baserunner.py]: good point, fixed now

**Comment by ben-spiller** [comment]
Great thanks for the review Scott!

**Pull request merged**